src/emu-i386/kvm.c

```diff
@@ -698,75 +698,71 @@ int kvm_vm86(struct vm86_struct *info)
   return vm86_ret;
 }
 
-/* Emulate do_dpmi_control() using KVM */
+/* Emulate indirect_dpmi_transfer() using KVM */
 int kvm_dpmi(sigcontext_t *scp)
 {
   struct vm86_regs *regs;
-  int ret;
   unsigned int trapno;
 
   monitor->tss.esp0 = offsetof(struct monitor, regs) +
     offsetof(struct vm86_regs, es);
 
   regs = &monitor->regs;
-  do {
-    regs->eax = _eax;
-    regs->ebx = _ebx;
-    regs->ecx = _ecx;
-    regs->edx = _edx;
-    regs->esi = _esi;
-    regs->edi = _edi;
-    regs->ebp = _ebp;
-    regs->esp = _esp;
-    regs->eip = _eip;
-
-    regs->cs = _cs;
-    regs->__null_ds = _ds;
-    regs->__null_es = _es;
-    regs->ss = _ss;
-    regs->__null_fs = _fs;
-    regs->__null_gs = _gs;
-
-    regs->eflags = _eflags;
-    regs->eflags &= (SAFE_MASK | X86_EFLAGS_VIF | X86_EFLAGS_VIP);
-    regs->eflags |= X86_EFLAGS_FIXED | X86_EFLAGS_IF;
 
-    kvm_run(regs);
+  regs->eax = _eax;
+  regs->ebx = _ebx;
+  regs->ecx = _ecx;
+  regs->edx = _edx;
+  regs->esi = _esi;
+  regs->edi = _edi;
+  regs->ebp = _ebp;
+  regs->esp = _esp;
+  regs->eip = _eip;
+
+  regs->cs = _cs;
+  regs->__null_ds = _ds;
+  regs->__null_es = _es;
+  regs->ss = _ss;
+  regs->__null_fs = _fs;
+  regs->__null_gs = _gs;
+
+  regs->eflags = _eflags;
+  regs->eflags &= (SAFE_MASK | X86_EFLAGS_VIF | X86_EFLAGS_VIP);
+  regs->eflags |= X86_EFLAGS_FIXED | X86_EFLAGS_IF;
 
-    /* orig_eax >> 16 = exception number */
-    /* orig_eax & 0xffff = error code */
-    trapno = regs->orig_eax >> 16;
+  kvm_run(regs);
 
-    _eax = regs->eax;
-    _ebx = regs->ebx;
-    _ecx = regs->ecx;
-    _edx = regs->edx;
-    _esi = regs->esi;
-    _edi = regs->edi;
-    _ebp = regs->ebp;
-    _esp = regs->esp;
-    _eip = regs->eip;
-
-    _cs = regs->cs;
-    _ds = regs->__null_ds;
-    _es = regs->__null_es;
-    _ss = regs->ss;
-    _fs = regs->__null_fs;
-    _gs = regs->__null_gs;
-
-    _eflags = regs->eflags;
-
-    ret = -1; /* mirroring sigio/sigalrm */
-    if (trapno != 0x20) {
-      _cr2 = (uintptr_t)MEM_BASE32(monitor->cr2);
-      _trapno = trapno;
-      _err = regs->orig_eax & 0xffff;
-      if (_trapno == 0x0e && vga_emu_fault(scp, 1) == True)
-	ret = dpmi_check_return(scp);
-      else
-	ret = dpmi_fault(scp);
-    }
-  } while (!ret);
-  /* "-3" means to call dpmi_fault1 in dpmi.c:_dpmi_control */
-  return ret == -3 ? 0 : ret;
+  /* orig_eax >> 16 = exception number */
+  /* orig_eax & 0xffff = error code */
+  trapno = regs->orig_eax >> 16;
+
+  _eax = regs->eax;
+  _ebx = regs->ebx;
+  _ecx = regs->ecx;
+  _edx = regs->edx;
+  _esi = regs->esi;
+  _edi = regs->edi;
+  _ebp = regs->ebp;
+  _esp = regs->esp;
+  _eip = regs->eip;
+
+  _cs = regs->cs;
+  _ds = regs->__null_ds;
+  _es = regs->__null_es;
+  _ss = regs->ss;
+  _fs = regs->__null_fs;
+  _gs = regs->__null_gs;
+
+  _eflags = regs->eflags;
+
+  /* fpstate is kept inside KVM */
+  __fpstate = NULL;
```

Comment:
But __fpstate is allocated in dpmi_init()
via aligned_alloc() and I don't see if you
conditionalized it, so I guess this produces
the mem leak.
I don't think its necessary to use per-client
fpstate indeed, but still I wonder if it is good
idea to just never save it. For example if we
switch to vm86 (also using kvm), don't we
need to save it and swap to the one of vm86?
At least that was the current logic.  Maybe
it is superfluous, but at least if the DPMI client
terminates, it would be surprising to not throw
away its FPU state and to keep using.

Reply:
And before you ask - no, there was no compelling
reason to switch from attribute to aligned_alloc(),
except some portability concerns (attributes are
always non-standard), and the fact that I find the
use of aligning attribute not very efficient inside
structure inside array (but yes, I know it is guaranteed
to work even in this case).

Reply:
No mem leak. This fpstate is part of the dpmi_stack_frame in nonnative_dpmi_transfer(), it does not overwrite DPMI_CLIENT.stack_frame's fpstate.

However you have a point: there is no need to allocate the fpstate any more since it is not used at all.

Switching between KVM vm86 and KVM dpmi already keeps the FPU state inside KVM. This switch only involves changing the VM bit in the eflags image on the stack and adjusting ssp0 in the TSS, FPU is not touched.

If, however, actual vm86() or cpuemu is used instead of KVM for V86 then yes we need state saving. However I haven't tested that and I suspect it doesn't work at all ($_cpu_vm="emulated" + $_cpu_vm_dpmi="kvm" -- there is no logic to deal with r/o JIT protection)

Reply:
> Switching between KVM vm86 and KVM dpmi already keeps the FPU state inside KVM.

Yes, but is this normal?
Esp if DPMI client exited (could also exit because
of some FPU exception after corrupting its fpstate).
The previous logic was to use separate FPU context
for vm86 IIRC.

> ($_cpu_vm="emulated" + $_cpu_vm_dpmi="kvm" -- there is no logic to deal with r/o JIT
> protection)

Maybe at least kvm+vm86() could work.
Also I think dealing with jit protection is rather
easy these days. I rewrote all the convoluted
code that did that.
In either case, if you think reusing single fpstate
is fine for the moment, lets just merge that and
keep a note for later.

Reply:
But conditionalizing the fpstate allocation would still be good.

Reply:
ok, will write a patch to do that.

Reply:
> ok, will write a patch to do that.

By the way, I wonder if we actually need
this all. coroutines are unhelpful for KVM.
And while they won't hurt now, imagine we
want to split DPMI into a different process
for the use with seccomp. Then coroutine,
if remains at all, will be moved to child. In
this case this patch will have to be reverted.

So maybe just to conditionalize the coroutine
creation?

Reply:
Ok. I pushed a new patchset without coroutines but with cleanups that make the debug log for KVM and native DPMI much more similar.

The last patch is a bit of a grab bag, it could be split into 5 small patches that do just one thing.